src/overcooked_ai_py/visualization/extract_events.py

```diff
@@ -0,0 +1,90 @@
+import numpy as np
+import itertools, copy
+from collections import defaultdict
+from overcooked_ai_py.utils import numpy_to_native
+def extract_events(trajectories, traj_idx=0, add_cumulative_events=True):
+    """
+    extracts events from trajectories for plots
+    """
+
+    def find_last_event(events, event_type, p, last_timestep=None):
+        for event in reversed(events):
+            if event.get("player") == p and event.get("action") == event_type \
+                    and (last_timestep is None or event["timestep"] < last_timestep):
+                return event
+            
+    def holding_event(events, p, t, obj):
+        last_pickup_t = (find_last_event(events, "pickup", p, t) or {}).get("timestep")
+
+        start_t = last_pickup_t or 0
+        event = {
+            "player": p,
+            "start_timestep":start_t,
+            "end_timestep":t,
+            "action": "holding",
+            "object": obj
+        }
+        return event
+
+    def get_holding_events(events, game_states):
+        holding_events = [holding_event(events, e.get("player"), e.get("timestep"), e.get("object")) 
+            for e in events if e.get("action") in ["drop", "potting", "delivery"]]
+        # check if object is held at end of the episode
+        for p, player in enumerate(game_states[-1].players):
+            if player.held_object:
+                holding_events.append(holding_event(events, p, len(game_states)-1, player.held_object.to_dict()))
+        return holding_events
+
+
+    def get_cumulative_events(events, last_timestep):
+        """
+        Receives events for scatter plot and returns events for cumulative plot
+        """
+        def add_cumulative_event(events, s, timestep, action, player=None, obj=None):
+            event = {"action": action,
+                    "sum": s,
+                    "timestep": timestep,}
+            if player is not None: event["player"] = player
+            if obj is not None: event["object"] = obj
+            events.append(event)
+
+        cumulative_events = []
+
+        s = 0
+        player_sums = defaultdict(int)
+        a = "pickups_and_drops"
+        
+        # add zero timestep events
+        first_timestep = 0
+        add_cumulative_event(cumulative_events, s, first_timestep, a)
+        for p in set(map(lambda x: x["player"], events)):
+            add_cumulative_event(cumulative_events, s, first_timestep, a, p)
+
+        for e in filter(lambda x:x["action"] in ["pickup", "drop"], events):
+            p = e["player"]
+            t = e["timestep"]
+            s+=1
```

Comment:
> Overall looks good. I left some pretty nitpicky feedback but all the requested changes should be very minor, if anything at all. As a general comment, I would be very careful with the object_id field -- I like the idea of adding a unique id to each game object, but I can forsee a lot of pitfalls (i.e. a serialized and then deserialized object not keeping its uid) as well as logic pitfalls, as the rest of Overcooked_mdp.py was written without the uid in mind. There might be cases where we are creating new objects that would have previously been considered equal instead of keeping the old object around. For example, before we used to create a new Onion object when it was added to a soup. I've long since changed that, but there might be other edge cases like that remaining that might lead to some subtle bugs in the future.

ObjectStates (and SoupStates) are serializing and deserializing with object_id. The only thing I see now is that I did not edited but should be are eq and hash methods (I will push code soon with them changed). There is also repr that could be changed, but I'm not sure if it is desirable.

IMO adding object_id elsewhere is even more bug-prone so adding it to ObjectStates is the least bad way to track a single object.

Reply:
done